### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by mottosso** [pyblish/pyblish-qml#108]
Memory keeps ticking upwards even during resets.

~10 publishes with ~20 plug-ins ends up at ~200mb of active memory use.

**Comment by mottosso** [comment]
@tokejepsen this is related to your findings in [this thread](http://forums.pyblish.com/t/slow-concurrent-publishes/118/11) but I still haven't been able to detect any performance decrease..

Is it reproducible on your end? Does it happen each time after a certain amount of publishes?

**Comment by tokejepsen** [comment]
Yeah, it happens everytime.

Do you have plugins that query something on the internet?
I wonder whether the plugins are being store in memory, so a single plugin that communicates with ftrack would get stored multiple times in memory, and then get executed multiple times.

**Comment by mottosso** [comment]
Plug-ins are reloaded at each turn and they are supposed to get garbage collected each time you refresh the plug-ins or start the GUI. But there's no way it can be generating megabytes of data at each turn.

You can test this by logging the value of a variable in module-scope.

```python
some_value = 0

class MyPlugin(pyblish.api.plugin):
  def process(self):
    global some_value
    some_value += 1
    self.log.info("The value is %i" % some_value)
```

It should always say "1", as it will get re-created at each refresh.

Either way, the `python.exe` process is only related to the GUI, it doesn't show you how much memory is occupied by plug-ins within Maya, where they are ultimately loaded and run.

I'm also pretty sure it's got to do with the messages and terminal. They might not get garbage collected inbetween runs. You could test it by doing something crazy, like this.

```python
import pyblish.api

class MyPlugin(pyblish.api.Validator):
    def process(self):
        for x in xrange(100):
            self.log.info("Message #%i" % x)

pyblish.api.register_plugin(MyPlugin)
```

This causes the GUI to take up an added 50 mb evertime the plug-in is run, and the memory doesn't go away, indicating that the messages aren't getting garbage collected.

**Comment by tokejepsen** [comment]
> This causes the GUI to take up an added 50 mb evertime the plug-in is run, and the memory doesn't go away, indicating that the messages aren't getting garbage collected.

We do have a fair bit of messages, but it seems strange that it would affect the performance like it does. Should I do a screen capture of the slowing down, or how do I show it?

**Comment by mottosso** [comment]
A screen capture would be great, something reproducible even better, like a script with one or more plug-ins responsible that I could register in-memory and run on my end.

**Comment by mottosso** [comment]
> but it seems strange that it would affect the performance like it does.

Yes and no, strange that it would have an effect on publishing, but the way it's implemented makes this a hazard. For each pair of instance/plug-in that gets triggered for processing, it's messages are first drawn in the GUI. It has to wait for the drawing to complete before it can continue processing, so if it's having trouble drawing, it will affect publishing performance.

**Comment by tokejepsen** [comment]
You might need to wait for the higher quality processing, but here is a screen recording; https://youtu.be/6UigjBURxJg

**Comment by mottosso** [comment]
That is amazingly slow!

Could you try this for me?

```python
import pyblish.util
pyblish.util.publish()
```

And tell me how long it takes?

**Comment by tokejepsen** [comment]
Very quick!

Used this code to get some data;
```python
import time
start_time = time.time()
import pyblish.util
pyblish.util.publish()
print("--- %s seconds ---" % (time.time() - start_time))
```

Can't say that the data is absolute, as the printing of the time is before the pyblish prints in the terminal, but it should give a relative overview. Basically it stayed around the same time for publishing.

8.44200015068
7.90900015831
6.96800017357
6.92699980736
7.18099999428
7.01899981499
7.13900017738

**Comment by mottosso** [comment]
Ok, well that limits the possibilities to the GUI, so that's good.

I'm still not seeing any slowdown on my end though, that's what's bothering me. Until this get's sorted, you could whip up a simple gui with a single button for them to do a "Quick publish", and stick to using Pyblish QML for deeper inspection into trouble.

Psuedo code

```python
from pyside.QtGui import QPushButton

p = QPushButton("Quick Publish")
p.clicked.connect(pyblish.util.publish)
p.show()
```

**Comment by tokejepsen** [comment]
Yeah, I might look at that.

Could it be something with having qml on a network? That's the only difference between you setup and ours.

**Comment by mottosso** [comment]
It might be slower in general on the network, if the network is slower than the local drive, but it shouldn't be affected by the number of times you publish.

**Comment by tokejepsen** [comment]
I have been working on the Maya pipeline here, and found that the slow down doesn't seem to appear here. It might be a bit premature as its not in production yet.
@mkolar, where did you find the slow downs? Nuke?

**Comment by mottosso** [comment]
That's very interesting, do keep an eye on that. I've only been testing this in Maya so far.

**Comment by mkolar** [comment]
We've had slow-downs in houdini and nuke. I have to check properly though to be able to give more details.

**Comment by mkolar** [comment]
> Might be worth adding that, normally, publishing is expected to occur less than 10 times a day from any software, averaging at around 2-3 depending on crunch. It is also expected that workstations are either logged off/shut down at the end of each day.

Just to give you some more numbers. We are averaging at around 8-10 publishes a day, in crunch we can easily hit 15-20 (simple comps based on templates, animation combined from library of movements etc.) Machines stay on most of the time. As soon it's idle for 10 minutes it runs rendering slave and stays like that overnight. So very long uptime unless artist restarts in the morning

**Comment by mottosso** [comment]
> Just to give you some more numbers.

Thanks for that @mkolar!

**Comment by BigRoy** [comment]
This still seems to be an active issue, especially since we're using a background process continuously running again at our studio (tray or shell in background). With bigger scenes where instances hold a lot of data the memory usage of pyblish can grow (and hold) over a gigabyte of memory. Even on smaller scenes the increase is around 10 MB per reset in the GUI.

When memory gets big in size the pyblish process also tends to lock up resulting in the GUI opening completely white (and staying like that) as if it can't update or the GUI not opening at all and the host giving a connection error.

**Comment by mottosso** [comment]
> pyblish can grow quickly to over a gigabyte of memory

Really!?

That's a problem. I've been thinking about this issue, and I'm fairly confident it's related to how items are implemented. They each have their own signalling mechanisms, as an alternative to calling the model to update each item. The idea was to simplify things, and make Python more like QML, but it hasn't really paid off and this can easily be killed off.

Here's how it works today.

```python
my_instance = self.item_model.instances[0]
my_instance.isToggled = False
```

This triggers the appropriate signal for the GUI to pick this up. It isn't standard Qt, but rather something I developed thinking I was being smart.

The standard way for that to happen is something closer to this.

```python
my_instance = self.item_model.instances["myInstance"]
my_index = self.item_model.instances.index(my_instance)
model_index = self.item_model.createIndex(my_index)
self.item_model.setData(model_index, False, self.item_model.IS_TOGGLED)
```

It's quite a bit more complex, which I why I went the other way, but at the end of the day it's much more efficient and less error prone.

I think that because each item handles its own signals, they aren't properly getting picked up by garbage collection during reset, which would explain the memory leak.
